### DISCUSSION TITLE
Resolved: `deprecated` does not work as expected

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#1275] [Ideas]
During the crate flattening work we have been deprecating the moved modules, types, and functions. I've been doing all this without testing (bad Tobin) - turns out `deprecated` does not work as we (I) expected and is basically useless how we are using it. It only works when annotating a function directly i.e., it does NOT work for modules, types, or their methods.

I would have expected the following snippet to throw at least 5 if not 6 warnings - it throws 1 (excl. unused import).
```rust
use std::str::FromStr;

use bitcoin::util::address::{Error, Address};
use bitcoin::util::bip158;
use bitcoin::util::misc::{BITCOIN_SIGNED_MSG_PREFIX, script_find_and_remove};

fn main() {
    let a = Address::from_str("1QJVDzdqb1VpbDK7uDeyVXy9mR27CJiyhY").unwrap();
    if a.is_standard() {
        //
    }
}
```

Only `script_find_and_remove` gets a warning. I think this means that we have to forget about deprecating moved modules/types, should we just announce the move in release notes instead?

(Discussion probably depends on resolution of this comment also: https://github.com/rust-bitcoin/rust-bitcoin/pull/1260#issuecomment-1247352653)

**Comment by Kixunil** [comment]
Oh, crap.

For types we can reexport with `type`, for functions we have to copy signatures and call the new ones.

**Comment by apoelstra** [comment]
Crap indeed.

Unfortunately I think we need useful deprecation warnings here, because users are gonna want useful information about how to port their code (and hopefully they'll be able to it in easy-to-review stages). So as @Kixunil says we're probably gonna have to implement all the functions in their old locations, calling the new ones.

**Comment by tcharding** [comment]
Fun times :)

**Comment by tcharding** [comment]
Unless I'm being brain dead there is no good way to move types. We can put a type alias or a `pub use` in the place where the type used to be but there is no way to indicate to downstream users that they need to use the new path. Unless we just leave the re-export/alias lying around for ever which seems like kruft we don't need at version sub 1.0

**Reply by Kixunil** [reply]
I believe `deprecated` on `type` works, doesn't it?

**Reply by tcharding** [reply]
Not for me it doesn't, I did try it.

**Reply by Kixunil** [reply]
In that case, I suggest to reexport, put a big warning into release notes and remove them in the following release.

**Reply by tcharding** [reply]
Sounds good, will do this.

**Comment by tcharding** [comment]
This has been reported a bunch of times: https://github.com/rust-lang/rust/issues/85388
